Add Feedback Prompt in Cart & Checkout blocks sidebar

Aljullu opened the pull request:
Fixes #1300.

I couldn't get it to appear after the _Advanced_ section. We could achieve this with some custom CSS using flexbox, but I'm not happy modifying the Gutenberg sidebar styles because we might break unexpected things now or in the future, so I would prefer another solution if anybody knows it or leaving the current implementation if we consider it's good enough.

#### Accessibility

- [x] I've tested using only a keyboard (no mouse)
- [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

![imatge](https://user-images.githubusercontent.com/3616980/70520770-59fe9400-1b3e-11ea-8aea-3f0fbff420a1.png)

### How to test the changes in this Pull Request:

1. Create a post with a _Cart_ or _Checkout_ blocks.
2. Verify there is a _Feedback?_ prompt in the sidebar.

senadir commented:
> I couldn't get it to appear after the Advanced section.

You can use `Slot` and `Fill` since the advanced section has a slot to use.
References: https://developer.wordpress.org/block-editor/components/slot-fill/
the slot name is `InspectorAdvancedControls`, it will show up before the class field and I'm not sure how to flip them without CSS

```js
<Fill name="InspectorAdvancedControls">
    Your content
</Fill>
```

nerrad commented:
We should use an approach similar to the custom_css setting so that this feedback panel can be enhanced on our blocks automatically given the attribute listing support for it: https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/hooks/custom-class-name.js

nerrad commented on assets/js/blocks/cart-checkout/feedback-prompt/index.js:
```diff
@@ -0,0 +1,44 @@
+			</h2>
+			<p className="wc-block-feedback-prompt__text">{ text }</p>
+			<a
+				href="https://wordpress.org/support/plugin/woo-gutenberg-products-block/reviews/"
```
Have we verified this is where we want feedback to go?

Aljullu replied:
AFAIK it's not decided yet, so I added this link as a placeholder. I was planning to ask @garymurray when he comes back from AFK. For now, I added a `@todo` comment so we don't forget about it (b88391af4edfc8ff36bb72ac538162ad174de145).

Aljullu commented:
> You can use `Slot` and `Fill` since the advanced section has a slot to use.
References: https://developer.wordpress.org/block-editor/components/slot-fill/
the slot name is `InspectorAdvancedControls`, it will show up before the class field and I'm not sure how to flip them without CSS, but there could be a possibility since both of them uses the slot
> 
> ```js
> <Fill name="InspectorAdvancedControls">
>     Your content
> </Fill>
> ```

Thanks for the tip @senadir! I was looking at it but it seems to add it inside the _Advanced_ panel, right? We would like to have it at the root level but I couldn't find any slot fill for it (at least, not [here](https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/components/block-inspector/index.js)).

> We should use an approach similar to the custom_css setting so that this feedback panel can be enhanced on our blocks automatically given the attribute listing support for it: https://github.com/WordPress/gutenberg/blob/master/packages/block-editor/src/hooks/custom-class-name.js

I converted it to a HOC (718d3f2cc0258be5c767866b61a5cbb24aea2555), and refactored it to use a filter (96bba029d61a383eafa2c0a1c08f7988e319b50d). I tried to copy the custom class inspector implementation as close as possible, but couldn't rely on the [`supports`](https://developer.wordpress.org/block-editor/developers/block-api/block-registration/#supports-optional) attribute because we need a specific copy for each block, so feedback is welcome.

nerrad commented on assets/js/hocs/with-feedback-prompt/index.js:
```diff
@@ -0,0 +1,51 @@
+ */
+const withFeedbackPrompt = createHigherOrderComponent( ( BlockEdit ) => {
+	return ( props ) => {
+		const feedbackPromptText = blocksFeedback[ props.name ];
```
This is a neat approach but I'm wary of having all the customization for feedback prompt in this one file.  It'd be more discoverable and easier to work with if individual blocks defined their specific elements.  What about one of these options?

### Option 1: specific named props for feedback prompt content and link.

In this option, edit components for blocks would be fed specific props that get picked up automatically by this HOC.  They'd need to be namespaced appropriately.

- `wcFeedbackContent`: For the custom content.
- `wcFeedbackUrl`: For the url where people submit their feedback.

### Option 2: use React.Context

This would be similar to option 2 but would utilize react context instead and the `withFeedbackPrompt` hoc would just read the values from the context.  Edit blocks would then just explicitly wrap anything receiving feedback with the feedback prompt provider:

_edit_
```jsx
<FeedbackProvider value={ feedBackConfig }>
    <EditComponent>
</FeedbackProvider>
```

_withFeedbackPrompt_:

```jsx
const withFeedbackPrompt = createHigherOrderComponent( ( BlockEdit ) => {
	return ( props ) => {
        const { content, url } = useFeedbackContext();
        if ( content ) { 
            /** do stuff **/
```

On the surface it seems like the context would be a bit more work, but the advantage is that the `EditComponent` or anything else in its composition doesn't have to care or worry about passing through any of the props and I like the explicitness in using the provider which makes it clear that this is a component with a feedback element.  I wonder if the Provider itself could take care of adding the `editor.BlockEdit` filter so everything is contained?

Aljullu replied:
I was playing with this idea but I'm not sure it's possible. The `editor.BlockEdit` is applied on an upper level than the block `edit` function is called, so wrapping the `<Edit>` block with a context provider, doesn't make that context available when calling the `editor.BlockEdit` filter.

I pushed a WIP for the checkout block here: https://github.com/woocommerce/woocommerce-gutenberg-products-block/pull/1356/commits/21f889b021ceea6fef722efab9663799829bc769. Something similar would happen with props (option 1): props passed in the block `edit` function are not available in the `editor.BlockEdit`.

Please, let me know if I'm missing something!

***

To solve the issue you raised:

> The only thing I'm wary about is having all the customization for feedback prompt in this one file (which is right now just content). It'd be more discoverable and easier to work with if individual blocks defined their specific elements. 

If the options above don't work, I thought about some alternative solutions:
* If we don't use the filter but rely only on the HOC like in 718d3f2cc0258be5c767866b61a5cbb24aea2555, we would have access to props/context defined in the `edit` function.
* We could alternatively create a `blockFeedbackRegistry` similar to the `blocksRegistry` we currently have in place. That would allow us to call a `registerBlockFeedback()` function right after `registerBlockType()` and have access to the `blockFeedbackRegistry` inside the `editor.BlockEdit` filter.

Thoughts?

nerrad commented:
Oh I also forgot to mention in my review, you need to export the new hoc from the hocs entry point (index.js) along with the other hocs.  I couldn't see the new feedback prompt until that was done.

Aljullu commented:
> Oh I also forgot to mention in my review, you need to export the new hoc from the hocs entry point (index.js) along with the other hocs. I couldn't see the new feedback prompt until that was done.

My bad, I probably removed it while cleaning the code before pushing. :man_facepalming:  Fixed in 3bd62dd9c6a9d96a68742fa8e6d01c2a083c19ab.

Aljullu commented:
Thanks for all the good ideas in this thread @nerrad! I updated this PR according to your feedback. Feel free to take another look.

> so it receives the content and url

For now, I left the URL hardcoded into the `<FeedbackPrompt>` component until we know if the feedback URL is going to be shared by all blocks or each one will have a different one.

nerrad commented on assets/js/hocs/with-feedback-prompt/index.js:
```diff
@@ -11,41 +10,25 @@ import { createHigherOrderComponent } from '@wordpress/compose';
-	};
-}, 'withFeedbackPrompt' );
+const withFeedbackPrompt = createHigherOrderComponent(
+	( content ) => ( BlockEdit ) => {
```
I think this should be:

```js
const withFeedbackPrompt = ( content ) => createHigherOrderComponent( /**/ );
```

In other words, the curried function should return the higher order component.

Aljullu replied:
Yes... I wasn't sure about this. Changed in: d016e8ff997d7bf23a961e2f4d3addec24c07a93.

garymurray commented:
We can now link this to the Ideas board - I have created a category for WooCommerce Blocks - this link should pre-filter it: https://ideas.woocommerce.com/forums/133476-woocommerce?category_id=384565

We can then also use that link for any future/past blocks that we want to capture feedback on.
